### ISSUE TITLE
UndoManager: incorrect undo result  (extra element in array after undo)

### ISSUE THREAD

**Issue by obeah** [y-crdt/y-crdt#380]
Hi, 

I hoped this test case is related to the same problem that was resolved in #371, but it seems to be a separate issue.

```rust
        let d = Doc::new();
        let r = d.get_or_insert_map("r");
        let s1 = r.insert(&mut d.transact_mut(), "s1", MapPrelim::<i32>::new());
        let b1_arr = s1.insert(&mut d.transact_mut(), "b1", ArrayPrelim::default());

        let b1_el1 = b1_arr.insert(&mut d.transact_mut(), 0, MapPrelim::<i32>::new());
        let b2_arr = b1_el1.insert(&mut d.transact_mut(), "b2", ArrayPrelim::default());
        let b2_arr_nest = b2_arr.insert(&mut d.transact_mut(), 0, ArrayPrelim::default());
        b2_arr_nest.insert(&mut d.transact_mut(), 0, 232291652);
        b2_arr_nest.insert(&mut d.transact_mut(), 1, -30);

        println!("Before undo manager: {}", r.to_json(&d.transact()));

        let mut mgr = UndoManager::with_options(&d, &r, Options::default());

        let mut txn = d.transact_mut();
        b2_arr_nest.remove(&mut txn, 1);
        b2_arr_nest.insert(&mut txn, 1, -5);

        drop(txn);
        mgr.reset();
        println!("After txn 1: {}", r.to_json(&d.transact()));

        let mut txn = d.transact_mut();

        let b1_el0 = b1_arr.insert(&mut txn, 0, MapPrelim::<i32>::new());
        let b2_0_arr = b1_el0.insert(&mut txn, "b2", ArrayPrelim::default());
        let b2_0_arr_nest = b2_0_arr.insert(&mut txn, 0, ArrayPrelim::default());
        b2_0_arr_nest.insert(&mut txn, 0, 232291652);
        b2_0_arr_nest.insert(&mut txn, 1, -6);
        b1_arr.remove(&mut txn, 1);

        let b1_el1 = b1_arr.insert(&mut txn, 1, MapPrelim::<i32>::new());
        b1_el1.insert(&mut txn, "f2", "C1");

        drop(txn);
        mgr.reset();
        println!("After txn 2: {}", r.to_json(&d.transact()));

        println!("Before undo sequence: {}", r.to_json(&d.transact()));
        mgr.undo().unwrap();
        println!("After undo 1: {}", r.to_json(&d.transact()));
        mgr.undo().unwrap();
        println!("After undo 2: {}", r.to_json(&d.transact()));

```
The output:

Before undo manager: {s1: {b1: [{b2: [[232291652, -30]]}]}}
After txn 1: {s1: {b1: [{b2: [[232291652, -5]]}]}}
After txn 2: {s1: {b1: [{b2: [[232291652, -6]]}, {f2: C1}]}}
Before undo sequence: {s1: {b1: [{b2: [[232291652, -6]]}, {f2: C1}]}}
After undo 1: {s1: {b1: [{b2: [[232291652, -5]]}]}}
After undo 2: {s1: {b1: [{b2: [[232291652, -30, -5]]}]}}

-5 is kept in the array after the last undo.
Important: **The problem occurs not on every run**

I have another case with more complex data structure which seem to be similar.
Initial state: {s1: {b-1: [{b-2: [{b-3: [{b-4: [-1448662, -2501]}, {b-4: [-13922, 15]}]}]}]}}
After 1st modification: {s1: {b-1: [{b-2: [{b-3: [{b-4: [-1448662, 11]}, {b-4: [-13922, 15]}]}]}]}}
After 2nd modification: {s1: {b-1: [{b-2: [{b-3: [{b-4: [-1448662, 11]}, {b-4: [-13922, 15]}]}, {b-3: [{b-4: [-14]}]}]}, {b-2: [{b-3: [{b-4: [-24]}]}]}]}}
1st undo: {"s1":{"b-1":[{"b-2":[{"b-3":[{"b-4":[-1448662.0,11.0]},{"b-4":[-13922.0,15.0]}]}]}]}}
2nd undo: {"s1":{"b-1":[{"b-2":[{"b-3":[{"b-4":[-1448662.0,-2501.0,11.0]},{"b-4":[-13922.0,15.0]}]}]}]}}

Again 1 extra element in "b-4". 
As in previous test case the problem occurs not on every run

**Comment by sinkingsugar** [comment]
I can confirm we have the same issue on our game creation system state crdt. It's making feel like we need to implement our own library cos it occurs randomly after deleting (remove()) items from maps. You experience that easily with undo/redo but even normal removes causes it.
The issue lies in the fact that sometimes delete sets data is no sent properly during diffs exchange.

**Comment by Horusiath** [comment]
I'm working on this issue currently. The reason why it does happen sometimes, seems to be related to a fact that we compute undo/redo items by storing them on a HashSet, which can have different order of elements depending on a run. However this is not a root cause, as changing it to BTreeSet doesn't fully solve the issue.

> The issue lies in the fact that sometimes delete sets data is no sent properly during diffs exchange.

AFAIK delete set is always sent properly, what I assume is the issue here is that delete set is not produced correctly as the result of bunked undo/redo operation. Correct me if I'm wrong (we didn't reached invalid updates that could cause desynchronising of remote replicas by years now, so I'm pretty confident on that part).

**Comment by sinkingsugar** [comment]
> AFAIK delete set is always sent properly, what I assume is the issue here is that delete set is not produced correctly as the result of bunked undo/redo operation. Correct me if I'm wrong (we didn't reached invalid updates that could cause desynchronising of remote replicas by years now, so I'm pretty confident on that part).

Yes the delete set is there, just not correct. But I'm not super sure. When the issue happens clients state vectors are correct (when compared) but snapshots diverge. I guess that's normal cos snapshots include the delete sets.
But yes it looks like it's the issue you are working on. But we definitely can cause it even with simple remove operations here.

I have a repro if needed but it's a bit bulky (it's a whole game state in our engine 🤣)

Given what you said I assume some delete operations are happening before inserts. Requiring a causality barrier somehow as a solution maybe?

Anyways thanks for looking into it!

**Comment by Horusiath** [comment]
We'll release the fix as part of v0.18.1
